File: MovieMatch-Frontend/interfaces/movie.ts
import { FormProps } from "@/components/Form";
import { User } from "@/constants";

export interface MovieSummary {
  id: number;
  title: string;
  releasedOn: Date;
  director: string;
  description: string;
  status: string;
  genres: GenreSummary[];
  picture: string;
  watchCount: number;
  rating: number;
}

export interface GenreSummary {
  id: number;
  name: string;
  photo: string;
}

export interface Genre extends GenreSummary {
  movies: MovieSummary[];
}

export interface MovieSection {
  title: string;
  movies: MovieSummary[];
}

export interface Review {
  id: number;
  comment: string;
  rating: number;
  user: User;
}

export interface Movie extends MovieSummary {
  user: User;
  reviews: Review[];
}

export type DashboardData = MovieSection[];

export interface MovieFormData
  extends Omit<
    MovieSummary,
    "genres" | "picture" | "user" | "id" | "rating" | "watchCount" | "status"
  > {
  genres: number[];
  picture: File | undefined;
}

export const movieForm: FormProps<MovieFormData, GenreSummary> = {
  button: {
    text: "Create",
  },
  fields: [
    {
      label: "Title",
      name: "title",
      type: "text",
    },
    {
      label: "Director",
      name: "director",
      type: "text",
    },
    {
      label: "Description",
      input: "textarea",
      name: "description",
      type: "text",
    },
    {
      label: "Released On",
      name: "releasedOn",
      type: "date",
      onChange: (e) => {
        return new Date(e.target.value);
      },
    },
    {
      label: "Picture",
      name: "picture",
      type: "file",
      inputProps: {
        accept: "image/*",
      },
      messages: [
        {
          message: "Only image files are allowed",
          match: "typeMismatch",
        },
      ],
      onChange: (e) => {
        return e.target.files[0];
      },
    },
  ],
};

File: MovieMatch-Server/Models/Movie.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace MovieMatch.Models
{
    public class Movie
    {
        public int Id { get; set; }

        public string Title { get; set; } = null!;

        public DateTime ReleasedOn { get; set; }

        public string Director { get; set; } = null!;

        public string Description { get; set; } = null!;

        public virtual User User { get; set; } = null!;

        [JsonConverter(typeof(JsonStringEnumConverter))]
        public MovieStatus Status { get; set; } = MovieStatus.Waiting;
        public virtual List<Genre> Genres { get; set; } = new();
        [JsonConverter(typeof(PhotoUrlConverter))]
        public string Picture { get; set; } = null!;

        public virtual List<Review> Reviews { get; set; } = new();

        [NotMapped]
        public int WatchCount
        {
            get
            {
                return Reviews.Count;
            }
        }

        [NotMapped]
        public double Rating
        {
            get
            {
                if (WatchCount > 0)
                {
                    return Reviews.Average(r => r.Rating);
                }

                return 0;
            }
        }
    }
}

public enum MovieStatus
{
    Deleted, Approved, Waiting, Rejected
}